### ISSUE TITLE
What's difference between DocumentRootName and DatabaseRootName?

### ISSUE THREAD

**Issue by Thaina** [GoogleCloudPlatform/google-cloud-dotnet#1829]
https://github.com/GoogleCloudPlatform/google-cloud-dotnet/blob/master/apis/Google.Cloud.Firestore.V1Beta1/Google.Cloud.Firestore.V1Beta1/ResourceNames.cs

It seem both class has nearly the same code and property. And `DocumentRootName` was not used anywhere. Something seem wrong here

Also I was suspect that `DocumentRootName` and `DocumentPathName` and `AnyPathName` should contains "CollectionID" because the nature of FireStore system need CollectionID to be the parent of all documents. Hence the root of document should have collection name specified

**Comment by jskeet** [comment]
No, they're different as they represent different resource names. `DatabaseRootName` has a path of `projects/{project}/databases/{database}`; `DocumentRootName` has a path of `projects/{project}/databases/{database}/documents`.

It's possible that `AnyPathName` should take a `collectionId` rather than a `documentId`.

As it happens, none of these resource names are used at the moment - normally they'd be used in partial classes for the request/response protos to make them easier to use. I'm not quite sure why those properties aren't being generated at the moment...

**Comment by Thaina** [comment]
@jskeet I see. Thank you very much

But still I think `DocumentPathName` and `AnyPathName` both should take `collectionID` because every document must live under a collection and that collection could be deep in hierachy

Also I have seen at least some mention about each ResourceNames, such as `DocumentPathName` have comment on link below. But `DocumentRootName` is nowhere

https://github.com/GoogleCloudPlatform/google-cloud-dotnet/blob/b9c045cf36ea0932d6fa98b470f433b51ac0e59d/apis/Google.Cloud.Firestore/Google.Cloud.Firestore/FirestoreDb.cs#L63-L71

**Comment by jskeet** [comment]
Yes, there's a comment - but that's just a comment, and should be ignored for now. (It's not a doc comment; just a thought to myself about *potential* changes later.)

The way resource names work for open-ended paths would make it odd for `DocumentPathName` to have a collection ID, IMO. We'll take a closer look at the configuration to see why these resource names are being generated at all - they should either be used, or not generated. It may well be that we stop generating them entirely. For the moment, I suggest you ignore them. (We don't expect 99% of users to refer to anything in the low-level package at all.)

**Comment by Thaina** [comment]
@jskeet Still it need to be parsed I think. And currently now I was using low level package because the high level package from nuget has a problem on serializing dictionary that not a `Dictionary<string,object>`

**Comment by jskeet** [comment]
> Still it need to be parsed I think.

I don't know what you mean by that. The fact that the high-level library doesn't use any of these resource name types proves that you don't need to either.

Rather than persist with the low-level API, could I suggest that you build your own local copy of the high-level API and keep that in a local nuget repository? If you really can't do that, I'll look into creating a new beta release of the high-level API. At that point, would you be able to move to the high level API? Please be aware that any time I spend answering questions about the low-level API that we don't expect to be used is time taken away from other work.

**Comment by Thaina** [comment]
@jskeet It would not be the problem that your high level API has no bug from the start and that is the reason. I was rely on low level API as of now because it has no bug. I was testing it and it work more than your low level API

And you know I was making report with pull request so many days ago to fix that problem. But I have seen no update in nuget for 2 weeks so I think you would still fixing something and it would not finished. But I need to work with firestore in the meantime

While you said you would not expect anyone using it, it just more reliable as of now and I don't need to wait for you to fix it

As I said, if the high level API was worked from the start and I don't need to wait for your fix with unknown timeline like that I would already use it. If you not expose the low level I would already use REST API. I don't start use low level API you know? I try to use your high level API first. And then it crash. Whose fault is that I need to use low level API for 2 weeks? Or you are saying I should just use REST in the meantime?

**Comment by jskeet** [comment]
I merged a change to allow serialization of protobufs on January 3rd: https://github.com/GoogleCloudPlatform/google-cloud-dotnet/commit/82e04814473b10f5db1a7273b3fcca7e9c8166a2

I haven't built a NuGet package with that commit in, but as I say, you could easily do so yourself locally - you don't need to apply any code changes.

Note that I don't regard not serializing protos as a bug - it's just something we hadn't designed for. The library on NuGet reliably behaves as expected, just not as you wanted. The code in GitHub now should behave reliably *and* as you expected.

As I said, if you really don't want to build locally, I can check whether the Firestore team is happy with me creating another interim release.

**Comment by Thaina** [comment]
@jskeet I cannot build my own because I use CI pipeline on bitbucket that would use `dotnet restore` to pull from nuget before publish it to the aws lambda server

If it not in nuget then I have only option is to scram all the repo into my project

**Comment by jskeet** [comment]
Right, that's all you needed to say, rather than implying that there were still code changes needed due to bugs.

I'll check with the Firestore team whether they're happy to create a new release.

In the meantime, I'll leave this issue open as I think we need to check whether the resource names are desirable at all - but rest assured that you never *have* to use them.

**Comment by Thaina** [comment]
@jskeet I don't said this issue relate to the bug. I just see that while the low level was being public and also reliable, it has ambiguous property like these class

And the bug from weeks ago is not just **not serializing protos** at that time it cannot serialize any dictionary that not attach `IDictionary<string,object>`

And I was doing investigation again. It came from `SerializeMap` function that was used by WriteBatch

https://github.com/GoogleCloudPlatform/google-cloud-dotnet/blob/3cc3a1b31bc1cf827deb243d115f8d8456cade13/apis/Google.Cloud.Firestore/Google.Cloud.Firestore/WriteBatch.cs#L59

And it seem like it was not fixed yet

https://github.com/GoogleCloudPlatform/google-cloud-dotnet/blob/82e04814473b10f5db1a7273b3fcca7e9c8166a2/apis/Google.Cloud.Firestore/Google.Cloud.Firestore/ValueSerializer.cs#L122-L136

**Comment by jskeet** [comment]
Okay, I'm not sure I understand the issue - please file a separate clear issue about this. In that issue, please provide an example using the *public* API, with expected results vs actual results. I'm very happy to try to improve the library, but the requests need to be clear and specific. It's always worth filing an issue rather than just a pull request, too, so that it's clear what you're trying to achieve rather than just one possible implementation of that.

**Comment by Thaina** [comment]
OK @jskeet 

**First** - This issue. I just file a question that these documents on the repo seem not complete. It just minor suspicious that if you say it would not be used anywhere then it's OK that it should be removed sometimes later and I agree with it. I just use it currently because it work like a path parser and state keeper

**Second** - You try to convince me to not use low level API. This is whole other problem that not relate to this issue. You just throw it out, not me

But then I need to state that there was so many problem so I can't do it. I was writing with low level API for 2 weeks now since the bug I reported. And you said you would request a publishing to nuget, that great but it was not about the problem

**Third** - I just remember that the problem I encounter 2 weeks ago is not just **cannot serialize Dictionary of protobuf Value**. The reason I was start using low level API was because I **cannot serialize any dictionary** at all. I try to write Newtonsoft's JObject

But at that time I don't know what really a problem so I try to do many things and end up with trying to serialize the dictionary of `Value` manually. And it still not work so I make a pull request. And I don't want to do reflection so I just make pull request just like that

But now I was knowing more about the library and so I was investigate again and found out that the actual problem came from the link I post above

So. How many issue I should make?

**Comment by jskeet** [comment]
Yes, I raised the high level API in this issue - because if we can fix everything that's preventing you from using the high level API, you won't need to worry about things like this issue. I will always prioritize issues with the high level API over questions around the low level API because they will be useful to more customers.

It sounds like a single new issue is all that's required, explaining what you want *your* code to look like. Note that we wouldn't take a direct dependency on Json.NET - there's already an indirect one, but we're hoping to get rid of that over time. So while I wouldn't expect serializing a `JObject` to work necessarily, it seems reasonable to want to serialize an `IDictionary<string, Foo>` where `Foo` is any type that can normally be serialized as a value. I'll start working on the code for that - if you could file the issue so I can make sure that the change satisfies it, that would be appreciated.

Basically I'm very keen on getting you to the stage where you can use the high level API. You gave a "thumbs up" to my previous pull request, so I assumed that was the only change required. Please feel free to file issues about anything that's blocking you from using the high level API, as it may well affect others too. I can't guarantee that we can satisfy everything directly, but we'll see what we can do.
